src/cmap/connection.ts

```diff
@@ -416,35 +421,10 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
       ? new Msg(cmdNs, finalCmd, commandOptions)
       : new Query(cmdNs, finalCmd, commandOptions);
 
-    const commandResponseHandler = inTransaction
-      ? (err?: AnyError, ...args: Document[]) => {
-          // We need to add a TransientTransactionError errorLabel, as stated in the transaction spec.
-          if (
-            err &&
-            err instanceof MongoNetworkError &&
-            !err.hasErrorLabel('TransientTransactionError')
-          ) {
-            err.addErrorLabel('TransientTransactionError');
-          }
-
-          if (
-            session &&
-            !cmd.commitTransaction &&
-            err &&
-            err instanceof MongoError &&
-            err.hasErrorLabel('TransientTransactionError')
-          ) {
-            session.transaction.unpinServer();
-          }
-
-          return callback(err, ...args);
-        }
-      : callback;
-
     try {
-      write(this, message, commandOptions, commandResponseHandler);
+      write(this, message, commandOptions, callback);
```

Comment:
where did we land with the TransientTransactionError label? Is it just no longer required?

Reply:
Hm, thanks for calling this out. I made a perhaps debatable decision to always make local network/timeout errors include this error label [here](https://github.com/mongodb/node-mongodb-native/blob/NODE-3011/load-balancer-support/src/error.ts#L562-L563). The transactions spec [states](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#transienttransactionerror): 

> In the case of command errors, the server adds the label; in the case of network errors or server selection errors where the client receives no server reply, the client adds the label.

I believe this used to be specific to the context of being in a transaction, but may be more important now for load balancers (and thus applied universally). A quick check in the dev chat did not elucidate much concern about including it always, except that libmongoc does indeed only apply it during transactions. It also didn't break any tests :shrug: Perhaps we should make a NODE ticket for it to track in the future, or even better a DRIVERS ticket to clarify if that language is missing an "when encountered in an active transaction" clause

Reply:
@mbroadst Looking over the transactions spec, it does appear to [define the transient transaction error](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#id16) as 

> Any command error that includes the "TransientTransactionError" error label in the "errorLabels" field. Any network error encountered running any command other than commitTransaction **in a transaction**. 

and it has the same "in a transaction" language in the sentence immediately preceding the one you quoted above:

> Any network error or server selection error encountered running any command besides commitTransaction **in a transaction**.

so to me at least it doesn't read like we should be adding that logic to all network errors in general outside of transactions, though I'm not sure I fully understand the implications for load balancers here

Reply:
My concern is that the current implementation adds the label to all network/timeout errors, regardless of load balancer mode, so even if we say the load balancer should always behave as being in a transaction of sorts, the same isn't true of non-load balancer mode. As you said, it is at best misleading and at worst breaking in some edge case that we don't test, though if we think the risk of the latter is low enough, it may be something we can live with, depending on how difficult it would be to implement a more correct solution.

Reply:
What I like about this solution is we remove one callback from every operation in a transaction, and make the code in the CMAP module a little easier to understand. I _think_ it might be enough to add the logic to [the server error handling code](https://github.com/mongodb/node-mongodb-native/blob/4.0/src/sdam/server.ts#L450) since we still execute all operations against a `Server` (rather than a `Connection`).

Reply:
I've moved this out now over the the server class error handling to check if we are in a transaction and apply the label and everything is passing.